File: gateways/js/src/agentid.js
import { Performative } from './performative.js';
import { ParameterReq  } from './message.js';
import { Gateway } from './gateway.js';  // import Gateway class for type checking. Remove if not needed.
import { Message } from './message.js';  // import Message class for type checking. Remove if not needed.

const DEFAULT_TIMEOUT = 10000; // Default timeout for non-owned AgentIDs


/**
* An identifier for an agent or a topic. This can be to send, receive messages, and set or get parameters
* on an agent or topic on the fjåge master container.
*/
export class AgentID {
  /**
   * Name of the agent
   * @type {string}
   */
  name;

  /**
   * Whether the identifer is for a topic
   * @type {boolean}
   */
  topic;

  /**
   * Gateway owner for this AgentID
   * @type {Gateway}
   */
  owner;

  /**
   * Construct an AgentID.
   *
   * @param {string} name
   * @param {boolean} [topic=false]
   * @param {Gateway} [owner]
   */
  constructor(name, topic=false, owner) {
    this.name = name;
    this.topic = topic;
    this.owner = owner;
    this._timeout = owner ? owner._timeout : DEFAULT_TIMEOUT; // Default timeout if owner is not provided
  }

  /**
  * Gets the name of the agent or topic.
  *
  * @returns {string} - name of agent or topic
  */
  getName() {
    return this.name;
  }

  /**
  * Returns true if the agent id represents a topic.
  *
  * @returns {boolean} - true if the agent id represents a topic, false if it represents an agent
  */
  isTopic() {
    return this.topic;
  }

  /**
  * Sends a message to the agent represented by this id.
  *
  * @param {Message} msg - message to send
  * @returns {void}
  */
  send(msg) {
    msg.recipient = this;
    if (this.owner) this.owner.send(msg);
    else throw new Error('Unowned AgentID cannot send messages');
  }

  /**
  * Sends a request to the agent represented by this id and waits for a reponse.
  *
  * @param {Message} msg - request to send
  * @param {number} [timeout=owner.timeout] - timeout in milliseconds
  * @returns {Promise<Message>} - response
  */
  async request(msg, timeout=this._timeout) {
    msg.recipient = this;
    if (this.owner) return this.owner.request(msg, timeout);
    else throw new Error('Unowned AgentID cannot send messages');
  }

  /**
  * Gets a string representation of the agent id.
  *
  * @returns {string} - string representation of the agent id
  */
  toString() {
    return this.toJSON() + ((this.owner && this.owner.connector) ? ` on ${this.owner.connector.url}` : '');
  }

  /**
  * Gets a JSON string representation of the agent id.
  *
  * @returns {string} - JSON string representation of the agent id
  */
  toJSON() {
    return (this.topic ? '#' : '') + this.name;
  }

  /**
   * Inflate the AgentID from a JSON string.
   *
   * @param {string} json - JSON string to be converted to an AgentID
   * @param {Gateway} [owner] - Gateway owner for this AgentID
   * @returns {AgentID} - AgentID created from the JSON string or object
   */
  static fromJSON(json, owner) {
    if (typeof json !== 'string') {
      throw new Error('Invalid JSON for AgentID');
    }
    json = json.trim();
    if (json.startsWith('#')) {
      return new AgentID(json.substring(1), true, owner);
    } else {
      return new AgentID(json, false, owner);
    }
  }

  /**
   * @overload
   * @param {string} param - parameter name to set
   * @param {unknown} value - parameter value to set
   * @param {number} [index=-1] - index of parameter to be set
   * @param {number} [timeout=owner.timeout] - timeout for the response
   * @returns {Promise<unknown>} - a promise which returns the new value of the parameter
   */

  /**
   * @overload
   * @param {string[]} params - parameter names to set
   * @param {unknown[]} values - parameter values to set
   * @param {number} [index=-1] - index of parameters to be set
   * @param {number} [timeout=owner.timeout] - timeout for the response
   * @returns {Promise<unknown[]>} - a promise which returns the new value of the parameters
   */

  /**
  * Sets parameter(s) on the Agent referred to by this AgentID.
  *
  * @param {(string|string[])} params - parameters name(s) to be set
  * @param {(unknown|unknown[])} values - parameters value(s) to be set
  * @param {number} [index=-1] - index of parameter(s) to be set
  * @param {number} [timeout=owner.timeout] - timeout for the response
  * @returns {Promise<(unknown | unknown[])>} - a promise which returns the new value(s) of the parameters
  */
  async set (params, values, index=-1, timeout=this._timeout) {
    if (!params) return null;
    let msg = new ParameterReq();
    msg.recipient = this;

    if (Array.isArray(params)) {
      if (!Array.isArray(values)) {
        throw new Error('Values must be an array because an array was passed for parameters');
      }

      if (params.length != values.length) throw new Error(`Parameters and values arrays must have the same length: ${params.length} != ${values.length}`);
      const clonedParams = params.slice(); // Clone the array to avoid side effects
      const clonedValues = values.slice(); // Clone the values array
      msg.param = clonedParams.shift();
      msg.value = clonedValues.shift();
      msg.requests = clonedParams.map((p, i) => {
        return {
          'param': p,
          'value': clonedValues[i]
        };
      });
    } else {
      msg.param = params;
      msg.value = values;
    }
    msg.index = Number.isInteger(index) ? index : -1;
    const rsp = await this.owner.request(msg, timeout);
    var ret = Array.isArray(params) ? new Array(params.length).fill(null) : null;
    if (!rsp || rsp.perf != Performative.INFORM || !('param' in rsp) || !rsp.param || typeof rsp.param !== 'string') {
      if (this.owner._returnNullOnFailedResponse) return ret;
      else throw new Error(`Unable to set ${this.name}.${params} to ${values}`);
    }
    if (Array.isArray(params)) {
      const values = 'values' in rsp && rsp.values ? rsp.values : {};
      values[rsp.param] = 'value' in rsp ? rsp.value : undefined;
      const rkeys = Object.keys(values);
      return params.map( p => {
        if (p.includes('.')) p = p.split('.').pop();
        let f = rkeys.find(k => (k.includes('.') ? k.split('.').pop() : k) == p);
        return f ? values[f] : undefined;
      });
    } else {
      return 'value' in rsp ? rsp.value : undefined;
    }
  }


  /**
   * @overload
   * @param {string} param - name of the parameter to get
   * @param {number} [index=-1] - index of the parameter to get
   * @param {number} [timeout=owner.timeout] - timeout for the response
   * @returns {Promise<unknown>} - a promise which returns the value of the parameter
   */

  /**
   * @overload
   * @param {string[] | null} param - names of the parameters to get, or null to get the value of all parameters on the Agent
   * @param {number} [index=-1] - index of parameters to get
   * @param {number} [timeout=owner.timeout] - timeout for the response
   * @returns {Promise<unknown[]>} - a promise which returns the values of the parameters
   */

  /**
  * Gets parameter(s) on the Agent referred to by this AgentID.
  *
  * @param {string | string[] | null} params - parameters name(s) to be get, null implies get value of all parameters on the Agent
  * @param {number} [index=-1] - index of parameter(s) to be get
  * @param {number} [timeout=owner.timeout] - timeout for the response
  * @returns {Promise<unknown | unknown[]>} - a promise which returns the value(s) of the parameters
  */
  async get(params, index=-1, timeout=this._timeout) {
    let msg = new ParameterReq();
    msg.recipient = this;
    if (params){
      if (Array.isArray(params)) {
        const clonedParams = params.slice(); // Clone the array to avoid side effects
        msg.param = clonedParams.shift();
        msg.requests = clonedParams.map(p => {return {'param': p};});
      }
      else msg.param = params;
    }
    msg.index = Number.isInteger(index) ? index : -1;
    const rsp = await this.owner.request(msg, timeout);
    var ret = Array.isArray(params) ? new Array(params.length).fill(null) : null;
    if (!rsp || rsp.perf != Performative.INFORM || !('param' in rsp) || !rsp.param || typeof rsp.param !== 'string') {
      if (this.owner._returnNullOnFailedResponse) return ret;
      else throw new Error(`Unable to get ${this.name}.${params}`);
    }
    // Request for listing of all parameters.
    if (!params) {
      const values = 'values' in rsp && rsp.values ? rsp.values : {};
      values[rsp.param] = 'value' in rsp ? rsp.value : undefined;
      return values;
    } else if (Array.isArray(params)) {
      const values = 'values' in rsp && rsp.values ? rsp.values : {};
      values[rsp.param] = 'value' in rsp ? rsp.value : undefined;
      const rkeys = Object.keys(values);
      return params.map( p => {
        if (p.includes('.')) p = p.split('.').pop();
        let f = rkeys.find(k => (k.includes('.') ? k.split('.').pop() : k) == p);
        return f ? values[f] : undefined;
      });
    } else {
      return 'value' in rsp ? rsp.value : undefined;
    }
  }
}

File: gateways/python/fjagepy/AgentID.py
from __future__ import annotations

import logging
from typing import Optional, Self, Any

logger = logging.getLogger(__name__)
logger.addHandler(logging.NullHandler())

# Default timeout for non-owned AgentIDs (in milliseconds)
DEFAULT_TIMEOUT = 10000

class AgentID:
    """An identifier for an agent or a topic. This can be used to send, receive messages,
    and set or get parameters on an agent or topic on the fjåge container.

    Args:
        name : name of the agent
        topic : True if this represents a topic. Defaults to False.
        owner : Gateway owner for this AgentID. Defaults to None.
    """

    def __init__(self, name: str, topic: bool = False, owner = None) -> None:
        if not isinstance(name, str) or not name:
            raise ValueError("AgentID name must be a non-empty string")
        self.name = name
        self.topic = topic
        self.owner = owner
        self.index = -1  # for indexed parameters
        self._timeout = owner._timeout if owner else DEFAULT_TIMEOUT

    def get_name(self) -> str:
        """Gets the name of the agent or topic.

        Returns:
            str: name of agent or topic
        """
        return self.name

    def is_topic(self) -> bool:
        """Returns True if the agent id represents a topic.

        Returns:
            bool: True if the agent id represents a topic, False if it represents an agent
        """
        return self.topic

    def send(self, msg:Message) -> None:
        """Sends a message to the agent represented by this id.

        Args:
            msg: message to send

        Raises:
            RuntimeError: if this AgentID has no owner (unowned AgentID cannot send messages)
        """
        msg.recipient = self
        if self.owner:
            self.owner.send(msg)
        else:
            raise RuntimeError('Unowned AgentID cannot send messages')

    def request(self, msg:Message, timeout: Optional[int] = None) -> Optional[Message]:
        """Sends a request to the agent represented by this id and waits for a response.

        Args:
            msg: request to send
            timeout (int, optional): timeout in milliseconds. Defaults to owner's timeout.

        Returns:
            Response message

        Raises:
            RuntimeError: if this AgentID has no owner (unowned AgentID cannot send messages)
        """
        if timeout is None:
            timeout = self._timeout
        msg.recipient = self
        if self.owner:
            return self.owner.request(msg, timeout)
        else:
            raise RuntimeError('Unowned AgentID cannot send messages')

    def get(self, index: Optional[int] = -1) -> dict[str, Any]:
        """Gets the values of all parameters on the agent.

        Args:
            index : index for indexed parameters. Defaults to -1 (no index).

        Returns:
            dict: dictionary of all parameters and their values

        Raises:
            RuntimeError: if this AgentID has no owner (unowned AgentID cannot get parameters)
        """
        from .Message import ParameterReq

        rsp = self.request(ParameterReq(index=index))
        if rsp is None or 'param' not in rsp.__dict__ or 'value' not in rsp.__dict__:
            return {}

        # the first parameter is in rsp.param and rsp.value and the others are in the dict rsp.values,
        # we combine them into a single dictionary
        params = {}
        if 'param' in rsp.__dict__ and 'value' in rsp.__dict__:
            params[rsp.param] = rsp.value
        if 'values' in rsp.__dict__ and isinstance(rsp.values, dict):
            params.update(rsp.values)
        return params

    def to_json(self) -> str:
        """Gets a JSON string representation of the agent id.

        Returns:
            str: JSON string representation of the agent id

        :meta private:
        """
        return ('#' if self.topic else '') + self.name

    @staticmethod
    def from_json(json_str: str, owner = None) -> "AgentID":
        """Inflate the AgentID from a JSON string.

        Args:
            json_str : JSON string to be converted to an AgentID
            owner: Gateway owner for this AgentID. Defaults to None.

        Returns:
            AgentID: AgentID created from the JSON string

        :meta private:
        """

        json_str = json_str.strip()
        if json_str.startswith('#'):
            return AgentID(json_str[1:], topic=True, owner=owner)
        else:
            return AgentID(json_str, topic=False, owner=owner)

    def __eq__(self, other) -> bool:
        if not isinstance(other, AgentID):
            return False
        return (self.name == other.name) and (self.topic == other.topic)

    def __hash__(self) -> int:
        return hash(self.to_json())

    def __str__(self) -> str:
        """Gets a string representation of the agent id.

        Returns:
            str: string representation of the agent id
        """
        owner_str = f"Gateway({self.owner.connector.host}:{self.owner.connector.port})" if self.owner else 'none'
        return f"AgentID(name={self.name}, topic={self.topic}, owner={owner_str})"

    def _repr_pretty_(self, p, cycle) -> None:
        """Pretty print support for IPython/Jupyter."""
        if (self.owner is None) or (not self.owner.is_connected()):
            p.text(str(self) if not cycle else '...')
            return

        # Print a Java style Agent information
        from .Message import ParameterReq
        rsp = self.request(ParameterReq(index=self.index))
        if rsp is None or 'param' not in rsp.__dict__ or 'value' not in rsp.__dict__:
            p.text(str(self) if not cycle else '...')
            return

        # the first parameter is in rsp.param and rsp.value and the others are in the dict rsp.values,
        # we combine them into a single dictionary
        params = {}
        if 'param' in rsp.__dict__ and 'value' in rsp.__dict__:
            params[rsp.param] = rsp.value
        if 'values' in rsp.__dict__ and isinstance(rsp.values, dict):
            params.update(rsp.values)

        if 'title' in params:
            p.text('<<< ' + str(params['title']) + ' >>>\n')
        else:
            p.text('<<< ' + str(self.name).upper() + ' >>>\n')
        if 'description' in params:
            p.text('\n' + str(params['description']) + '\n')

        # First we take all the parameters and sort them alphabetically
        param_names = sorted(params.keys())

        # Then we split them into lists based on the prefix. Everything before the
        # final dot is considered a section.
        sections = {}
        for param in param_names:
            if '.' in param:
                section, subparam = param.rsplit('.', 1)
            elif param == 'title' or param == 'description':
                continue
            else:
                section, subparam = '', param
            if section not in sections:
                sections[section] = []
            sections[section].append((subparam, params[param], param))


        for section in sorted(sections.keys()):
            if section:
                p.text(f'\n[{section}]\n')
            for subparam, value, param in sections[section]:
                readonly = 'readonly' in rsp.__dict__ and isinstance(rsp.__dict__["readonly"], list) and param in rsp.__dict__["readonly"]
                p.text(f'  {subparam} {"=>" if readonly else "="} {value}\n')

    ## Magic methods to support syntactic sugar

    def __lshift__(self, msg) -> Optional[Message]:
        """ Supports sending messages through the << operator.
            Example: agent << msg will send the message msg to the agent represented by agent."""

        return self.request(msg)

    def __getitem__(self, index) -> Self:
        """ Supports indexed parameter access through the [] operator.
            Example: agent[1].param will refers to the first indexed parameter "param" of the agent."""

        # make a copy of this AgentID with the specified index
        new_aid = AgentID(self.name, topic=self.topic, owner=self.owner)
        new_aid.index = index
        new_aid._timeout = self._timeout
        return new_aid

# Magic methods to support dynamic parameter access using dot notation

def __getter(self, param: str) -> None | Any:
    if param in ['name', 'owner', 'topic', 'index', '_timeout'] or param.startswith('_ipython'):
        return self.__dict__[param]

    from .Message import ParameterReq

    rsp = self.request(ParameterReq(index=self.index).get(param))
    if rsp is None or 'param' not in rsp.__dict__ or 'value' not in rsp.__dict__:
        return None
    return rsp.__dict__.get('value', None)


setattr(AgentID, '__getattr__', __getter)


def __setter(self, param : str, value : Any) -> Any | None:
    if param in ['name', 'owner', 'topic', 'index', '_timeout']:
        self.__dict__[param] = value
        return value

    from .Message import ParameterReq
    rsp = self.request(ParameterReq(index=self.index).set(param, value))
    if rsp is None or 'param' not in rsp.__dict__ or 'value' not in rsp.__dict__:
        return None
    return rsp.__dict__.get('value', None)

setattr(AgentID, '__setattr__', __setter)